### PR TITLE
task(metadata): add app and device metadata fields to fallback/windows notifier

### PR DESCRIPTION
**Pull request by rich-bugsnag**
## Goal

Add app and device metadata fields to fallback/windows notifier

## Changeset

- Added generic (data present on both windows and fallbacks) data collection to the App and Device constructors
- Added platform specific data collection (lowmemory, isLaunching) to the PopulateApp and PopulateDevice methods in the fallback and windows NativeClient classes
- Added a lowMemoryWarning listener to the fallback and windows native client classes
- Added MarkLaunchAsCompleted implementation to the fallback and windows native client classes

## Testing

Added Windows CI test

### CODE REVIEW

File: features/desktop/unhandled_errors.feature [hunk 1]
```diff
@@ -13,6 +13,50 @@ Feature: Reporting unhandled events
...
+        And the event "device.id" is not null
+        And the event "device.locale" is not null
+        And the event "device.manufacturer" is not null
+        And the event "device.model" is not null
+        And the event "device.osName" is not null
+        And the event "device.osVersion" is not null
+        And the event "device.runtimeVersions" is not null
+        And the event "device.screenDensity" is not null
+        And the event "device.screenResolution" is not null
+        And the event "device.time" is not null
```
Comment (fractalwrench) [inline features/desktop/unhandled_errors.feature:45]: Would it be possible to validate that this is a timestamp, and that other app/device fields are non-zero integers etc?

File: src/BugsnagUnity/Native/Fallback/NativeClient.cs [hunk 2]
```diff
@@ -12,18 +13,42 @@ class NativeClient : INativeClient
 
         public IDelivery Delivery { get; }
 
-        private Dictionary<string, object> _editorMetadata = new Dictionary<string, object>();
+        private Dictionary<string, object> _fallbackMetadata = new Dictionary<string, object>();
+
+        private bool _launchMarkedAsCompleted = false;
+
+        private bool _hasRecivedLowMemoryWarning = false;
```
Comment (fractalwrench) [inline src/BugsnagUnity/Native/Fallback/NativeClient.cs:20]:
```suggestion
        private bool _hasReceivedLowMemoryWarning = false;
```

File: src/BugsnagUnity/Native/Fallback/NativeClient.cs [hunk 3]
```diff
@@ -12,18 +13,42 @@ class NativeClient : INativeClient
...
 
         public NativeClient(IConfiguration configuration)
         {
             Configuration = configuration;
             Breadcrumbs = new Breadcrumbs(configuration);
             Delivery = new Delivery();
+            Application.lowMemory += () => { _hasRecivedLowMemoryWarning = true; };
```
Comment (fractalwrench) [inline src/BugsnagUnity/Native/Fallback/NativeClient.cs:27]: Is there any way of detecting when the app exits the lowMemory state, or is this not possible with the available APIs?

**Review by fractalwrench — APPROVED** [review]
LGTM pending CI & resolution of comments

File: src/BugsnagUnity/Native/Fallback/NativeClient.cs (hunk 2 above)
Comment (rich-bugsnag) [reply src/BugsnagUnity/Native/Fallback/NativeClient.cs]: thanks 👍

File: features/desktop/unhandled_errors.feature (hunk 1 above)
Comment (rich-bugsnag) [reply features/desktop/unhandled_errors.feature:45]: I wrote this following the example of other metadata tests. @twometresteve are there better ways i could be asserting here?

Comment (twometresteve) [reply features/desktop/unhandled_errors.feature:45]: We do have an existing step that the Android (and probably others) use:
```    
And the event "device.time" is a timestamp
```

This Cucumber step is defined in MazeRunner and I can show you the docs that we have for them (they're not perfect, but could well help).

For Fractal's other suggestions, I'm not sure that we already have steps defined - but we can easily add them.  Let's talk IRL about that!

File: src/BugsnagUnity/Native/Fallback/NativeClient.cs (hunk 3 above)
Comment (rich-bugsnag) [reply src/BugsnagUnity/Native/Fallback/NativeClient.cs:27]: Not possible unfortunatly

File: features/desktop/unhandled_errors.feature (hunk 1 above)
Comment (rich-bugsnag) [reply features/desktop/unhandled_errors.feature]: Added timestamp and non 0 int checks

**Pull request merged**